internal/pkg/agent/install/install.go

```diff
@@ -32,81 +32,134 @@ const (
+	err := os.MkdirAll(pathDir, 0755)
+	if err != nil {
+		return errors.New(err, fmt.Sprintf("failed to create directory (%s) for shell wrapper (%s)", pathDir, wrapperPath), errors.M("directory", pathDir))
+	}
+	// Install symlink for darwin instead of the wrapper script.
+	// Elastic-agent should be first process that launchd starts in order to be able to grant
+	// the Full-Disk Access (FDA) to the agent and it's child processes.
+	// This is specifically important for osquery FDA permissions at the moment.
+	if runtime.GOOS == darwin {
+		// Check if previous shell wrapper or symlink exists and remove it so it can be overwritten
+		if _, err := os.Lstat(wrapperPath); err == nil {
+			if err := os.Remove(wrapperPath); err != nil {
+				return errors.New(err, fmt.Sprintf("failed to remove (%s)", wrapperPath), errors.M("destination", wrapperPath))
 			}
-			pt.Describe(fmt.Sprintf("Successfully created user %s", username))
 		}
+		err = os.Symlink(filepath.Join(topPath, paths.BinaryName), wrapperPath)
+		if err != nil {
+			return errors.New(err, fmt.Sprintf("failed to create elastic-agent symlink (%s)", wrapperPath), errors.M("destination", wrapperPath))
+		}
+	} else {
+		// We use strings.Replace instead of fmt.Sprintf here because, with the
+		// latter, govet throws a false positive error here: "fmt.Sprintf call has
+		// arguments but no formatting directives".
+		shellWrapperFormatted := strings.Replace(paths.ShellWrapper, "%s", topPath, -1)
+		err = os.WriteFile(wrapperPath, []byte(shellWrapperFormatted), 0755)
+		if err != nil {
+			return errors.New(err, fmt.Sprintf("failed to write shell wrapper (%s)", wrapperPath), errors.M("destination", wrapperPath))
+		}
+	}
+	return nil
+}
+
+// perform pre-install tasks, including uninstalling the previous install if needed
+func setupInstall(pt *progressbar.ProgressBar, topPath string, cfgFile string, log *logp.Logger) (string, error) {
+	dir, err := findDirectory()
+	if err != nil {
+		return "", errors.New(err, "failed to discover the source directory for installation", errors.TypeFilesystem)
+	}
+
+	// We only uninstall Agent if it is currently installed.
+	status, _ := Status(topPath)
+	// Uninstall current installation
+	//
+	// There is no uninstall token for "install" command.
+	// Uninstall will fail on protected agent.
+	// The protected Agent will need to be uninstalled first before it can be installed.
+	if status == Installed {
+		pt.Describe("Uninstalling current Elastic Agent")
+		err = Uninstall(cfgFile, topPath, "", log, pt)
+		if err != nil {
+			return "", errors.New(err, fmt.Sprintf("failed to uninstall Agent at (%s)", filepath.Dir(topPath)), errors.M("directory", filepath.Dir(topPath)))
+		}
+		pt.Describe("Successfully uninstalled current Elastic Agent")
 	}
+	return dir, nil
+}
 
+// perform the file copy operations
+func copyFiles(pt *progressbar.ProgressBar, topPath string, streams *cli.IOStreams, sourceDir string, ownership utils.FileOwner) error {
```

Comment:
This function does too much: I understand that this is the result of a refactor that extracted a big chunk of the install code into its own function but we don't need a ProgressBar to perform a copy of the files.

As an example have a look at https://github.com/elastic/elastic-agent/issues/3890 (and even then the cli.IOStreams can be pulled out of the function)